third_party/accessibility/ax/platform/ax_platform_node_textrangeprovider_win.cc

```diff
@@ -0,0 +1,1698 @@
+  BASE_DCHECK(common_ancestor_platform_node);
+  AXPlatformNodeDelegate* common_ancestor_delegate =
+      common_ancestor_platform_node->GetDelegate();
+  BASE_DCHECK(common_ancestor_delegate);
+  const gfx::Rect text_range_container_frame_bounds =
+      common_ancestor_delegate->GetBoundsRect(AXCoordinateSystem::kFrame,
+                                              AXClippingBehavior::kUnclipped);
+  UIA_VALIDATE_BOUNDS(text_range_container_frame_bounds);
+
+  gfx::Point target_point;
+  if (align_to_top) {
+    target_point = gfx::Point(root_frame_bounds.x(), root_frame_bounds.y());
+  } else {
+    target_point =
+        gfx::Point(root_frame_bounds.x(),
+                   root_frame_bounds.y() + root_frame_bounds.height());
+  }
+
+  if ((align_to_top && start()->GetAnchor()->IsText()) ||
+      (!align_to_top && end()->GetAnchor()->IsText())) {
+    const gfx::Rect text_range_frame_bounds =
+        common_ancestor_delegate->GetInnerTextRangeBoundsRect(
+            start_common_ancestor->text_offset(),
+            end_common_ancestor->text_offset(), AXCoordinateSystem::kFrame,
+            AXClippingBehavior::kUnclipped);
+    UIA_VALIDATE_BOUNDS(text_range_frame_bounds);
+
+    if (align_to_top) {
+      target_point.Offset(0, -(text_range_container_frame_bounds.height() -
+                               text_range_frame_bounds.height()));
+    } else {
+      target_point.Offset(0, -text_range_frame_bounds.height());
+    }
+  } else {
+    if (!align_to_top)
+      target_point.Offset(0, -text_range_container_frame_bounds.height());
+  }
+
+  const gfx::Rect root_screen_bounds = root_delegate->GetBoundsRect(
+      AXCoordinateSystem::kScreenDIPs, AXClippingBehavior::kUnclipped);
+  UIA_VALIDATE_BOUNDS(root_screen_bounds);
+  target_point += root_screen_bounds.OffsetFromOrigin();
+
+  AXActionData action_data;
+  action_data.action = ax::mojom::Action::kScrollToPoint;
+  action_data.target_node_id = common_ancestor_anchor->id();
+  action_data.target_point = target_point;
+  if (!common_ancestor_delegate->AccessibilityPerformAction(action_data))
+    return E_FAIL;
+  return S_OK;
+}
+
+// This function is expected to return a subset of the *direct* children of the
+// common ancestor node. The subset should only include the direct children
+// included - fully or partially - in the range.
+HRESULT AXPlatformNodeTextRangeProviderWin::GetChildren(SAFEARRAY** children) {
+  UIA_VALIDATE_TEXTRANGEPROVIDER_CALL_1_OUT(children);
+  std::vector<gfx::NativeViewAccessible> descendants;
+
+  AXPlatformNodeWin* start_anchor =
+      GetPlatformNodeFromAXNode(start()->GetAnchor());
+  AXPlatformNodeWin* end_anchor = GetPlatformNodeFromAXNode(end()->GetAnchor());
+  AXPlatformNodeWin* common_anchor = GetLowestAccessibleCommonPlatformNode();
+  if (!common_anchor || !start_anchor || !end_anchor)
+    return UIA_E_ELEMENTNOTAVAILABLE;
+
+  SAFEARRAY* safe_array = SafeArrayCreateVector(VT_UNKNOWN, 0, 0);
```

Comment:
Why is this different from the Chromium implementation?

Reply:
I'll add a TODO. This would require us to implement a method to get all children nodes that are between two particular nodes on the tree in traversal order, which I am not aware of currently having.